packages/wallet-sdk/src/provider/SolanaProvider.ts

```diff
@@ -0,0 +1,373 @@
+// Copyright (c) 2018-2022 Coinbase, Inc. <https://www.coinbase.com/>
+// Licensed under the Apache License, version 2.0
+
+import { ScopedLocalStorage } from "../lib/ScopedLocalStorage";
+import { RequestArguments } from "./Web3Provider";
+import SafeEventEmitter from "@metamask/safe-event-emitter";
+import { Transaction, PublicKey, SendOptions } from "@solana/web3.js";
+import { randomBytesHex } from "../util";
+import { SolanaWeb3Method } from "../relay/solana/SolanaWeb3Method";
+import { SolanaWeb3Response } from "../relay/solana/SolanaWeb3Response";
+export const SOLANA_PROVIDER_ID = "window.coinbaseSolana";
+
+type ErrorResponse = {
+  method: string;
+  message: string;
+};
+
+type SolanaWeb3Provider = {
+  publicKey: PublicKey | null;
+  isConnected: boolean;
+  connect: () => Promise<void>;
+  disconnect: () => Promise<void>;
+  signTransaction: (transaction: Transaction) => Promise<Transaction>;
+  signAllTransactions: (transactions: Transaction[]) => Promise<Transaction[]>;
+  signAndSendTransaction: (
+    transaction: Transaction,
+    options?: SendOptions,
+  ) => Promise<{ signature: string }>;
+  signMessage: (
+    message: Uint8Array,
+  ) => Promise<{ signature: Uint8Array } | Error>;
+  handleResponse: (event: any) => void;
+};
+
+export class SolanaProvider
+  extends SafeEventEmitter
+  implements SolanaWeb3Provider
+{
+  private eventManager = new Map<string, any>();
+  private storage: ScopedLocalStorage = new ScopedLocalStorage(
+    "coinbaseSolana",
+  );
+  isConnected: boolean;
+  publicKey: PublicKey | null;
+
+  public constructor() {
+    super();
+    this.connect = this.connect.bind(this);
+    this.sendTransaction = this.sendTransaction.bind(this);
+    this.signMessage = this.signMessage.bind(this);
+    this.signTransaction = this.signTransaction.bind(this);
+    this.request = this.request.bind(this);
+    this.isConnected = false;
+    this.publicKey = null;
+    window.addEventListener("message", this.handleResponse);
+  }
+
+  public handleResponse = (event: any) => {
+    if (event.data.type !== "extensionUIResponse") return;
+    const data = event.data.data;
+    const id = data.id;
+    const action = data.action;
+
+    const callback = this.getCallback(id);
+
+    if (callback) {
+      this.eventManager.delete(id);
+
+      switch (action) {
+        case SolanaWeb3Response.web3RequestCanceled:
+          callback(undefined, new Error("User canceled request"));
```

Comment:
Based on [the docs for this check](https://codeql.github.com/codeql-query-help/javascript/js-unvalidated-dynamic-method-call/) I _think_ this is fine because we're checking that we have a callback before entering the switch.